### PR TITLE
Apidocs

### PR DESCRIPTION
**Pull request by pault84**
api docmentation generation...

make docs will generate swagger.json. in cmd/osd/ might move that  i guess.

### CODE REVIEW

**Comment by pault84** [comment]
looks like i'm going to have to rebase

**Comment by pault84** [comment]
Alright so i have to figure out how to add examples to some of the parameters from volumeEnumerate and snapshotEnumerate.

swagger 2.0 does not allow for serialized objects as a form parameter.
which causes the {} part of the object to disappear and that breaks the api call.

i can also add examples.

guidance would be nice.

File: api/api.go [hunk 1]
```diff
@@ -80,77 +80,61 @@ const (
...
-	Id        string
-	Cpu       float64 // percentage.
-	MemTotal  uint64
-	MemUsed   uint64
-	MemFree   uint64
-	Avgload   int
-	Status    Status
+	// Id of the cluster.
+	Id string
+	// Id of the cluster.
```
Comment (adityadani) [inline api/api.go]: CpuUsage of the node in percentage.

File: api/api.go [hunk 2]
```diff
@@ -80,77 +80,61 @@ const (
...
+	Cpu float64 // percentage.
+	// Total Memory of the node
```
Comment (adityadani) [inline api/api.go]: Shouldnt this be of the form
MemTotal is the total memory of the node
The same applies to the rest of the fields as well.

File: api/api.go [hunk 3]
```diff
@@ -80,77 +80,61 @@ const (
...
+// swagger:model
 type Node struct {
-	Id        string
-	Cpu       float64 // percentage.
-	MemTotal  uint64
-	MemUsed   uint64
-	MemFree   uint64
-	Avgload   int
-	Status    Status
+	// Id of the cluster.
```
Comment (adityadani) [inline api/api.go]: Id of the node.

File: api/client/cluster/client.go [hunk 4]
```diff
@@ -142,94 +142,6 @@ func (c *clusterClient) EnableUpdates() error {
 	return nil
 }
 
-func (c *clusterClient) SetLoggingURL(loggingURL string) error {
```
Comment (adityadani) [inline api/client/cluster/client.go]: Mark it as deprecated. And we will define a strategy to deprecate this if not needed.

File: api/server/cluster.go [hunk 5]
```diff
@@ -57,6 +52,22 @@ func (c *clusterApi) String() string {
...
+// Lists cluster Nodes.
+//
+// This will return the entire cluster object and it's nodes.
+//
+// ---
+// produces:
+// - application/json
+// responses:
+//   '200':
+//      description: current cluster state
```
Comment (adityadani) [inline api/server/cluster.go]: Can we add description in the form of complete sentence as this is going to be a part of the swagger docs?

File: api/server/cluster.go [hunk 6]
```diff
@@ -316,6 +238,18 @@ func nodeStatusIntl() (api.Status, error) {
...
+// swagger:operation GET /cluster/nodestatus node status nodeStatus
+//
+// This will return the node status .
+//
+// ---
+// produces:
+// - application/json
+// responses:
+//   '200':
+//      description: node status
```
Comment (adityadani) [inline api/server/cluster.go]: returns the status of the node serving this http request.

File: api/server/cluster.go [hunk 7]
```diff
@@ -346,6 +292,18 @@ func (c *clusterApi) nodeHealth(w http.ResponseWriter, r *http.Request) {
...
+// This will return the peer node status
+//
+// ---
+// produces:
+// - application/json
+// responses:
+//   '200':
+//      description: anode
+//      schema:
+//         type: string
```
Comment (adityadani) [inline api/server/cluster.go]: this API takes in a nodeID as an argument

File: api/server/cluster.go [hunk 8]
```diff
@@ -426,6 +436,25 @@ func (c *clusterApi) versions(w http.ResponseWriter, r *http.Request) {
...
+// - name: resource
+//   in: path
+//   description: resourcetype to get alerts with
+//   required: true
+//   schema:
+//    "$ref": "#/definitions/ResourceType"
+// responses:
+//   '200':
+//      description: Alerts object
+//      schema:
```
Comment (adityadani) [inline api/server/cluster.go]: type: array

File: api/server/volume.go [hunk 9]
```diff
@@ -67,6 +67,30 @@ func (vd *volAPI) parseID(r *http.Request) (string, error) {
...
+//
+// Creates a single volume with given spec.
+//
+// ---
+// produces:
+// - application/json
+// parameters:
+// - name: spec
+//   in: body
+//   description: spec to create volume with
```
Comment (adityadani) [inline api/server/volume.go]: not really spec but it takes the VolumeCreateRequest. You could say "request to create volume"

File: api/server/volume.go [hunk 10]
```diff
@@ -531,6 +824,27 @@ func (vd *volAPI) quiesce(w http.ResponseWriter, r *http.Request) {
 	json.NewEncoder(w).Encode(volumeResponse)
 }
 
+// swagger:operation GET /osd-volumes/unquiesce/{id} volume unquiesce unquiesceVolume
```
Comment (adityadani) [inline api/server/volume.go]: this is POST

File: api/server/volume.go [hunk 11]
```diff
@@ -488,6 +760,27 @@ func (vd *volAPI) requests(w http.ResponseWriter, r *http.Request) {
 	json.NewEncoder(w).Encode(requests)
 }
 
+// swagger:operation GET /osd-volumes/quiesce/{id} volume quiesce quiesceVolume
```
Comment (adityadani) [inline api/server/volume.go]: this is POST

File: api/api.go (hunk 3 above)
Comment (pault84) [reply api/api.go]: good catch

File: api/api.go (hunk 1 above)
Comment (pault84) [reply api/api.go]: Even better catch (copy paste hero)

File: api/api.go (hunk 2 above)
Comment (pault84) [reply api/api.go]: You are right i'll change that

File: api/server/cluster.go (hunk 5 above)
Comment (pault84) [reply api/server/cluster.go]: Yes we can i just took what they used in the swagger documentation

File: api/server/cluster.go (hunk 7 above)
Comment (pault84) [reply api/server/cluster.go]: it does? i didn't see it in the route. but good catch!

File: api/server/cluster.go (hunk 8 above)
Comment (pault84) [reply api/server/cluster.go]: Alerts is an object that has an array in it.

File: api/client/cluster/client.go [hunk 12]
```diff
@@ -160,76 +161,6 @@ func (c *clusterClient) SetLoggingURL(loggingURL string) error {
 
 }
 
-func (c *clusterClient) SetManagementURL(managementURL string) error {
```
Comment (lpabon) [inline api/client/cluster/client.go]: Why is this removed for a docs PR? If this needs to be removed, please create a new PR just for this change.

Comment (pault84) [reply api/client/cluster/client.go]: None of these features have been used as far as i know but i can put them back and file a seperate pr for htose

**Comment by pault84** [comment]
this is good too merge.

**Comment by lpabon** [comment]
I'm going to review it tomorrow

**Comment by lpabon** [comment]
Looks like there are a lot of errors but we can fix them after this PR. Could you create an issue to fix the errors? See: https://app.swaggerhub.com/apis/libopenstorage/osd-api/1.0.0

**Comment by lpabon** [comment]
Two changes needed:

1. Could you add to the Makefile to automatically download go-swagger.
1. Output of swagger.json should not be in `cmd/osd`. Could we add it to `api/swagger` ?
